### PR TITLE
[KS-71] External peering core service

### PR DESCRIPTION
**Pull request by bolekk**
1. New "Capabilities" config, which includes Registry and Peering
2. PeerWrapper, which converts config and manages external Peer's lifecycle
3. Stub for RegistrySyncer

### CODE REVIEW

File: core/services/ocr2/delegate.go [hunk 1]
```diff
@@ -697,6 +703,11 @@ func (d *Delegate) newServicesGenericPlugin(
 		}
 		srvs = append(srvs, job.NewServiceAdapter(oracle))
 
+		if externalPeerWrapper != nil {
+			syncer := capabilities.NewSyncer(externalPeerWrapper.GetPeer(), capabilitiesRegistry, false, lggr)
```
Comment (cedric-cordenier) [inline core/services/ocr2/delegate.go:707]: @bolekk Not sure I understand why we're instantiating the syncer here. This means that for any generic plugin we'll instantiate a new syncer which doesn't seem quite right.

IMO the syncer should run outside of the job spawner as a core part of the application.

File: core/services/chainlink/application.go [hunk 2]
```diff
@@ -391,6 +392,12 @@ func NewApplication(opts ApplicationOpts) (Application, error) {
 		return nil, fmt.Errorf("P2P stack required for OCR or OCR2")
 	}
 
+	var externalPeerWrapper *externalp2p.ExternalPeerWrapper
+	if cfg.P2PExternal().Enabled() {
+		externalPeerWrapper = externalp2p.NewExternalPeerWrapper(keyStore.P2P(), cfg.P2PExternal(), db, globalLogger)
+		srvcs = append(srvcs, externalPeerWrapper)
```
Comment (cedric-cordenier) [inline core/services/chainlink/application.go:398]: @bolekk This is roughly where I was thinking we'd instantiate a syncer.

File: core/config/docs/core.toml [hunk 3]
```diff
@@ -434,6 +434,32 @@ DeltaReconcile = '1m' # Default
 # but the host and port must be fully specified and cannot be empty. You can specify `0.0.0.0` (IPv4) or `::` (IPv6) to listen on all interfaces, but that is not recommended.
 ListenAddresses = ['1.2.3.4:9999', '[a52d:0:a88:1274::abcd]:1337'] # Example
 
+[Capabilities.Registry]
+# Enabled enables Registry
+Enabled = false # Default
```
Comment (cedric-cordenier) [inline core/config/docs/core.toml:439]: @bolekk Does this enable the registry or just syncing? The registry should always be enabled IMO, but there's an argument to be made about disabling syncing.

File: core/services/chainlink/application.go [hunk 4]
```diff
@@ -185,7 +187,22 @@ func NewApplication(opts ApplicationOpts) (Application, error) {
 	keyStore := opts.KeyStore
 	restrictedHTTPClient := opts.RestrictedHTTPClient
 	unrestrictedHTTPClient := opts.UnrestrictedHTTPClient
-	registry := capabilities.NewRegistry(globalLogger)
+
+	var registry commontypes.CapabilitiesRegistry
+	if cfg.Capabilities().Registry().Enabled() {
+		registry = capabilities.NewRegistry(globalLogger)
```
Comment (cedric-cordenier) [inline core/services/chainlink/application.go:193]: Hmmmm... this is problematic: I think the capability registry should always be enabled as it's a core service.

File: core/capabilities/syncer.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		if err != nil {
+			return err
+		}
+		peers[p] = defaultStreamConfig
+	}
+	return s.peerWrapper.GetPeer().UpdateConnections(peers)
+}
+
+func (s *registrySyncer) Close() error {
+	return nil
```
Comment (krehermann) [inline core/capabilities/syncer.go:66]: should some connections be closed here?

File: core/capabilities/syncer.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/libocr/ragep2p"
+	ragetypes "github.com/smartcontractkit/libocr/ragep2p/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+)
+
+type registrySyncer struct {
```
Comment (krehermann) [inline core/capabilities/syncer.go:16]: can you add a comment about what the registry syncer does? it work be ideal the go doc explains every capability and what function it serves

File: core/config/capabilities_config.go [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package config
+
+type Capabilities interface {
+	Registry() CapabilityRegistry
+	Peering() P2P
+}
+
+type CapabilityRegistry interface {
+	Enabled() bool
```
Comment (krehermann) [inline core/config/capabilities_config.go:9]: this is suspicious. do you really need an interface only to tell you if a feature is on or off? instead can you write a help func that infers it from whether or not something concrete is nil?

somewhere later in the code you are using if registry != nil, which seems to mix level of abstractions

File: core/capabilities/syncer.go (hunk 5 above)
Comment (bolekk) [reply core/capabilities/syncer.go:66]: yeah we could leave it to the Peer but probably makes sense to call UpdateConnections() with an empty set here, to match Start(). I'll add it, thanks!

File: core/config/capabilities_config.go (hunk 7 above)
Comment (bolekk) [reply core/config/capabilities_config.go:9]: removing for now, as discussed offline.

File: core/config/docs/core.toml (hunk 3 above)
Comment (bolekk) [reply core/config/docs/core.toml:439]: I'll make it always-enabled.

**Review by cedric-cordenier — APPROVED** [review]

**Pull request merged**